### PR TITLE
bug fixes for missing meridional velocity extraction

### PR DESCRIPTION
**Pull request by jdha**
Addressing ticket #75

### CODE REVIEW

**Comment by jpolton** [comment]
@jdha just noticed that this is still in draft... Some comments anyway:
1) PyNEMO is hard to read, maybe all codes are. With an intention to help new users become new contributors, I think it would be a good idea to be liberal about adding one line comments about what is going on when we make edits.
2) I'm sure when this pull request is finished you will provide a test that must be passed. Even if it is a statement about the error that no longer stops the code.

**Comment by jdha** [comment]
@jpolton
re 2) I just need to verify the output before removing draft status
re 1) This is not something I'm concerned about at the moment - main focus is getting the benchmark working + tides, then getting meaningful CI and code coverage sorted before tidying. This is a long term goal - the class structure and subroutines need to be trimmed down. There are several routines in odd places, the extract routines are just a sprawling mess.

**Comment by jdha** [comment]
Tracer and 3D velocity extraction works [results to be verified]
Barotropic vels + SSH extraction missing a whole bunch of code.

Next steps:
Run pynemo for each ln_tra, ln_dyn2d, ln_dyn3d and combinations thereof

**Comment by jdha** [comment]
ln_tra, ln_dyn2d, ln_dyn3d run and complete individually - however, ln_dyn2d + ln_tra is broken
also need to remove the 3D vels write-out if only ln_dyn2d requested

File: pynemo/nemo_bdy_extr_tm3.py [hunk 1]
```diff
@@ -141,7 +145,7 @@ def __init__(self, setup, SourceCoord, DstCoord, Grid, var_nam, grd, pair):
         
         
         if self.key_vec:
-            self.nvar = self.nvar / 2
+            self.nvar = self.nvar // 2 # TODO: if  self.nvar=3 then this will pass and should not!
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:148]: Tried to find instances where I relied on integer division under py2.7

File: pynemo/nemo_bdy_extr_tm3.py [hunk 2]
```diff
@@ -87,10 +87,13 @@ def __init__(self, setup, SourceCoord, DstCoord, Grid, var_nam, grd, pair):
...
         sc_z = SC.zt[:]
         sc_z_len = len(sc_z)
-        
+      
         
         self.jpj, self.jpi = DC.lonlat[grd]['lon'].shape
         self.jpk = DC.depths[grd]['bdy_z'].shape[0]
+        self.bdy_dz= DC.depths[grd]['bdy_dz']
+        if grd=='t':
+            self.bdy_msk=DC.bdy_msk
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:96]: in the move to py3 the write-out of bdy_msk go lost. Popped it into the Extract object to carry over to the write subroutine

File: pynemo/nemo_bdy_extr_tm3.py [hunk 3]
```diff
@@ -322,23 +326,34 @@ def __init__(self, setup, SourceCoord, DstCoord, Grid, var_nam, grd, pair):
...
+        rep_dims_2d = (id_121.shape[0], id_121.shape[1], 1)
         # These tran/tiles work like matlab. Tested with same Data.
+        id_121_2d = id_121.repeat(1).reshape(rep_dims_2d).transpose(2, 0, 1)
+        reptile = np.arange(1).repeat(num_bdy).reshape(1, 
+                                                              num_bdy)
+        reptile = reptile.repeat(3).reshape(num_bdy, 3, 1, 
+                                            order='F').transpose(2, 0, 1)
+
+        id_121_2d = sub2ind((1, num_bdy), id_121_2d, reptile)
+
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:338]: It looks like in a effort to clean things up I did away with the 'z' grid and grouped all tracers and ssh under the 't' grid - this caused a few issues, so I have to carry around a 2d and 3d version of things. These are all quick fixes to get things working. I will remove the Matlab-style vectorisation in favour of loops and NUMBA in the next iteration.

File: pynemo/nemo_bdy_extr_tm3.py [hunk 4]
```diff
@@ -394,21 +409,29 @@ def __init__(self, setup, SourceCoord, DstCoord, Grid, var_nam, grd, pair):
...
 
         self.d_bdy = {}
+        
+        # Need to qualify for key_vec
         for v in range(self.nvar):
-            self.d_bdy[self.var_nam[v]] = {}
+            if grd == 'v':
+                 self.d_bdy[self.var_nam[v+1]] = {}
+            else:
+                 self.d_bdy[self.var_nam[v]] = {}
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:433]: for some reason in Tom's code either u or v is extracted - we need to keep both for the rotation onto the destination grid. u and v are always index 0 and 1 in Extract object - again a very quick and dirty fix. In future we should have a general Extract Class and use _inheritance_ for the different permutations.

File: pynemo/profile.py [hunk 5]
```diff
@@ -148,6 +148,7 @@ def process_bdy(setup_filepath=0, mask_gui=False):
     for grd in ['t', 'u', 'v']:
         DstCoord.depths[grd]['bdy_H']  = np.nanmax(z.zpoints['w'+grd], axis=0)
         DstCoord.depths[grd]['bdy_dz'] = np.diff(z.zpoints['w'+grd], axis=0)
+        DstCoord.depths[grd]['bdy_dz'] = np.vstack([DstCoord.depths[grd]['bdy_dz'], np.zeros((1,nbdy[grd]))])
```
Comment (jdha) [inline pynemo/profile.py:151]: need dz thicknesses to calculate depth integrated u/v

File: pynemo/nemo_bdy_ncgen.py [hunk 6]
```diff
@@ -59,6 +59,8 @@ def CreateBDYNetcdfFile(filename, N, I, J, K, rw, h, orig, fv, calendar, grd):
                                        fill_value=fv)
         varsalID = ncid.createVariable('vosaline', 'f4', ('time_counter', 'z', 'yb', 'xb', ),
                                        fill_value=fv)
+        varsshID = ncid.createVariable('sossheig', 'f4', ('time_counter', 'yb', 'xb', ),
+                                       fill_value=fv)
```
Comment (jdha) [inline pynemo/nemo_bdy_ncgen.py:63]: move ssh into grid T

File: pynemo/nemo_bdy_extr_tm3.py [hunk 7]
```diff
@@ -869,19 +967,32 @@ def write_out(self, year, month, ind, unit_origin):
...
             
-#        for v in self.variables:
-        for v in self.var_nam:
-            if self.settings['dyn2d']: # Calculate depth averaged velocity
+        #        for v in self.variables:
+        if self.key_vec == True:
+            if self.rot_dir == 'i':
+                varnams = [self.var_nam[0],]
+            else:
+                varnams = [self.var_nam[1],]
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:975]: as both source u and v are passed, use rot_dir to determine whether we're calculating destination u or destination v

File: pynemo/nemo_bdy_extr_tm3.py [hunk 8]
```diff
@@ -869,19 +967,32 @@ def write_out(self, year, month, ind, unit_origin):
...
+        else:
+            varnams=self.var_nam
+
+        for v in varnams:
+            if self.settings['dyn2d'] and ( (v == 'vozocrtx') or (v=='vomecrty') ) : # Calculate depth averaged velocity
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:980]: calc depth averaged velocity

File: pynemo/nemo_bdy_extr_tm3.py [hunk 9]
```diff
@@ -809,27 +896,38 @@ def time_interp(self, year, month):
         # Take the difference of the first two time enteries to get delta t
         
         del_t = time_counter[1] - time_counter[0]
-        dstep = 86400 / np.int(del_t)
+        dstep = 86400 // np.int(del_t)
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:899]: py3 syntax

File: pynemo/nemo_bdy_extr_tm3.py [hunk 10]
```diff
@@ -809,27 +896,38 @@ def time_interp(self, year, month):
...
         # multiple of 86400 | data are annual means
+        if self.key_vec == True:
+            if self.rot_dir == 'i':
+                varnams = [self.var_nam[0],]
+            else:
+                varnams = [self.var_nam[1],]
+        else:
+            varnams=self.var_nam
+
+
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:915]: determine whether we're using source u and v to calc destination u or destination v

File: pynemo/nemo_bdy_extr_tm3.py [hunk 11]
```diff
@@ -421,9 +444,15 @@ def extract_month(self, year, month):
         # Check year entry exists in d_bdy, if not create it.
         for v in range(self.nvar):
             try:
-                self.d_bdy[self.var_nam[v]][year]
+                if (self.key_vec == True and self.rot_dir == 'j'):
+                    self.d_bdy[self.var_nam[v+1]][year]
+                else:
+                    self.d_bdy[self.var_nam[v]][year]
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:450]: this is a fudge to set the varname to meridional velocity if rot_dir is j

File: pynemo/nemo_bdy_extr_tm3.py [hunk 12]
```diff
@@ -494,12 +523,15 @@ def extract_month(self, year, month):
...
-            n = self.nvar
+          #  n = self.nvar
 #            meta_data[n] = self.fnames_2[first_date].get_meta_data(self.var_nam[n], meta_data[n])
-            meta_data[n] = self.fnames_2.get_meta_data(self.var_nam[n], meta_data[n])
+            meta_data[1] = self.fnames_2.get_meta_data(self.var_nam[1], meta_data[1])
 
         for vn in range(self.nvar):
-            self.d_bdy[self.var_nam[vn]]['date'] = sc_time.date_counter[first_date:last_date + 1] 
+            if (self.key_vec == True and self.rot_dir == 'j'):
+                self.d_bdy[self.var_nam[vn+1]]['date'] = sc_time.date_counter[first_date:last_date + 1]
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:532]: again just correcting name if we're extracting meridional velocity

File: pynemo/nemo_bdy_extr_tm3.py [hunk 13]
```diff
@@ -508,9 +540,11 @@ def extract_month(self, year, month):
             #self.logger.info('opening nc file: %s', sc_time[f].file_name)            
             # Counters not implemented
 
-            sc_bdy = np.zeros((len(self.var_nam), sc_z_len, ind.shape[0], 
-                              ind.shape[1]))
+            sc_bdy={}
+           # sc_bdy = np.zeros((len(self.var_nam), sc_z_len, ind.shape[0], 
+           #                   ind.shape[1]))
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:545]: using a dict of np arrays now as we're mixing 2d and 3d variables

File: pynemo/nemo_bdy_extr_tm3.py [hunk 14]
```diff
@@ -600,32 +654,48 @@ def extract_month(self, year, month):
...
+                varid = sc_time[self.var_nam[vn]]
+                # Determine if slab or not
+                isslab = len(varid._get_dimensions())==3
+
+                # Set up tmp dict of tmp arrays
+                if isslab:
+                    sc_z_len=1
+                else:
+                    sc_z_len=self.sc_z_len
+
```
Comment (jdha) [inline pynemo/nemo_bdy_extr_tm3.py:670]: differentiate between 2D and 3D .... this naively assumes there is always a time dimension... something that is mostly true, but will have to be tidied up

File: plotting/plot_coords.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+
```
Comment (jpolton) [inline plotting/plot_coords.py:2]: Add a doc str

File: plotting/plot_bdy.py [hunk 16]
```diff
@@ -0,0 +1,275 @@
+#!/usr/bin/env python3
+
```
Comment (jpolton) [inline plotting/plot_bdy.py:2]: Add a doc string

File: pynemo/profile.py [hunk 17]
```diff
@@ -328,11 +328,16 @@ def process_bdy(setup_filepath=0, mask_gui=False):
                 
                 # Extract the data for a given month and year
                 
+                print('Extract')
+                print(key)
+                print(val)
                 extract_obj[key].extract_month(year, month)
```
[the remark this answers is not in the document]
Comment (jdha) [reply pynemo/profile.py]: already removed (note the Outdated badge)

File: plotting/plot_coords.py (hunk 15 above)
Comment (jdha) [reply plotting/plot_coords.py:2]: > @jdha All the changes look sensible by eye. Can you state which test I should conduct? I will also then check it's interaction with tide options (I expect no interaction). Then it is good to go.

All I've done so far is to run the benchmark extraction for ln_tra, ln_dyn2d and ln_dyn3d and combinations thereof. 
Output hasn't been verified yet. Plan is to generate a synthetic velocity field rotated at 45\deg and magnitude 2**0.5 to produce and u and v field of 1m/s - which could then be used as one of the unit tests.

**Review by jpolton — APPROVED** [review]

**Pull request merged**